### PR TITLE
CMake: cache TestBoostBug, clean up cached variables

### PR DESCRIPTION
**Pull request by tamiko**
This commit ensures that the test result of TestBoostBug is correctly
cached (similarly to all other compilation checks).

Further, make sure that all cached variables for compatibility checks for
ADOLC and Boost are correctly unset again.

### CODE REVIEW

**Comment by tamiko** [comment]
@masterleinad @davydden I think this small improvement (correctly caching the test results) makes the initial hit of the compatibility check much more amenable. Please test.

File: cmake/configure/configure_boost.cmake [hunk 1]
```diff
@@ -134,42 +132,50 @@ MACRO(FEATURE_BOOST_FIND_EXTERNAL var)
...
+        MESSAGE(STATUS "Performing Test BOOST_SERIALIZATION_USABLE")
+        IF(DEAL_II_ALLOW_PLATFORM_INTROSPECTION)
+          FILE(REMOVE_RECURSE ${CMAKE_CURRENT_BINARY_DIR}/cmake/configure/TestBoostBug)
+          FILE(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/cmake/configure/TestBoostBug)
+          EXECUTE_PROCESS(
+            COMMAND ${CMAKE_COMMAND}
+                      -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+                      -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+                      "-DBOOST_INCLUDE_DIRS=${BOOST_INCLUDE_DIRS}"
+                      "-DBOOST_LIBRARIES=${BOOST_LIBRARIES}"
```
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: @masterleinad The problem why you needed the workaround for the spearators were these missing quotes...

File: cmake/configure/configure_boost.cmake [hunk 2]
```diff
@@ -117,73 +117,79 @@ MACRO(FEATURE_BOOST_FIND_EXTERNAL var)
...
+        MESSAGE(STATUS "Performing Test BOOST_SERIALIZATION_USABLE")
+        IF(DEAL_II_ALLOW_PLATFORM_INTROSPECTION)
+          FILE(REMOVE_RECURSE ${CMAKE_CURRENT_BINARY_DIR}/cmake/configure/TestBoostBug)
+          FILE(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/cmake/configure/TestBoostBug)
+          EXECUTE_PROCESS(
+            COMMAND ${CMAKE_COMMAND}
+              -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+              -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+              "-DBOOST_INCLUDE_DIRS=${BOOST_INCLUDE_DIRS}"
+              "-DBOOST_LIBRARIES=${BOOST_LIBRARIES}"
```
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: @masterleinad The problem why you needed the workaround for the spearators were these missing quotes...

Comment (masterleinad) [reply cmake/configure/configure_boost.cmake]: OK, that's much easier indeed.

File: cmake/configure/configure_boost.cmake [hunk 3]
```diff
@@ -117,73 +117,79 @@ MACRO(FEATURE_BOOST_FIND_EXTERNAL var)
...
-          ERROR_QUIET
-        )
+    IF(DEAL_II_ALLOW_PLATFORM_INTROSPECTION)
+      #
+      # Test that Boost.Serialization is usable.
+      #
+      IF(NOT DEFINED BOOST_SERIALIZATION_USABLE OR NOT ${BOOST_SERIALIZATION_USABLE})
+        # Only run this check if it hasn't successfully run previously.
+        MESSAGE(STATUS "Performing Test BOOST_SERIALIZATION_USABLE")
+        IF(DEAL_II_ALLOW_PLATFORM_INTROSPECTION)
```
[the remark this answers is not in the document]
Comment (tamiko) [reply cmake/configure/configure_boost.cmake]: We did indeed. Fixing.

File: cmake/modules/FindBOOST.cmake [hunk 4]
```diff
@@ -103,4 +103,6 @@ DEAL_II_PACKAGE_HANDLE(BOOST
     Boost_LIBRARY_DIR_DEBUG Boost_LIBRARY_DIR_RELEASE
     _Boost_COMPONENTS_SEARCHED _Boost_INCLUDE_DIR_LAST
     _Boost_LIBRARY_DIR_LAST _Boost_USE_MULTITHREADED_LAST
+    BOOST_IOSTREAMS_USABLE # clean up check in configure_boost.cmake
+    BOOST_SERIALIZATION_USABLE # clean up check in configure_boost.cmake
   )
```
Comment (tamiko) [inline cmake/modules/FindBOOST.cmake]: @masterleinad 
It is a bit hidden in the `CONFIGURE_FEATURE` macro. All variables after `CLEAR` get unset if an external feature gets disabled. So, for something like external -> internal -> external, all checks should be rerun.

**Comment by davydden** [comment]
this PR worked for me on Ubuntu with patched 1.65.1, will check macOS now...

**Comment by davydden** [comment]
./run-tests

**Pull request merged**